extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/runtime/ConcurrencyLimiter.java

```diff
@@ -0,0 +1,102 @@
+package io.quarkus.websockets.next.runtime;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.websockets.next.WebSocketServerConnection;
+import io.smallrye.mutiny.helpers.queues.Queues;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+
+/**
+ * Used to limit concurrent invocations.
+ */
+class ConcurrencyLimiter {
+
+    private static final Logger LOG = Logger.getLogger(ConcurrencyLimiter.class);
+
+    private final WebSocketServerConnection connection;
+    private final Queue<Action> queue;
+    private final AtomicLong uncompleted;
+    private final AtomicLong queueCounter;
+
+    ConcurrencyLimiter(WebSocketServerConnection connection) {
+        this.connection = connection;
+        this.uncompleted = new AtomicLong();
+        this.queueCounter = new AtomicLong();
```

Comment:
In theory , if you need telemetry; you could just use the `uncompleted` one which can report the number of in-flight requests

Reply:
The `queueCounter` is currently only used to distinguish queued actions in log messages. So I don't think we can use `uncompleted` there. In theory, we could initialize the `queueCounter` only if `Logger.isDebugEnabled()`. However, I do remember some problems with native image as described in https://github.com/quarkusio/quarkus/issues/27735.

Reply:
https://github.com/quarkusio/quarkus/issues/39148